crates/runtime/src/instance.rs

```diff
@@ -65,13 +65,21 @@ pub struct Instance {
     ///
     /// This is where all runtime information about defined linear memories in
     /// this module lives.
-    memories: PrimaryMap<DefinedMemoryIndex, Memory>,
+    ///
+    /// The `MemoryAllocationIndex` was given from our `InstanceAllocator` and
+    /// must be given back to the instance allocator when deallocating each
+    /// memory.
+    memories: PrimaryMap<DefinedMemoryIndex, (MemoryAllocationIndex, Memory)>,
```

Comment:
Unsure whether this is better as-written, or if moving the `MemoryAllocationIndex` into `wasmtime_runtime::Memory` is better. Feel free to bike shed.